File: android/app/src/main/kotlin/com/edgelink/core/Pairing.kt
package com.edgelink.core

import java.security.MessageDigest
import java.util.Locale

data class SasCode(
    val numeric: String,
    val display: String = "${numeric.substring(0, 3)} ${numeric.substring(3, 6)}"
)

object Pairing {
    fun commitment(hostPublicKey: ByteArray, hostNonce: ByteArray): ByteArray =
        sha256(hostPublicKey, hostNonce)

    fun sas(
        hostPublicKey: ByteArray,
        clientPublicKey: ByteArray,
        hostNonce: ByteArray,
        clientNonce: ByteArray
    ): SasCode {
        val digest = sha256(hostPublicKey, clientPublicKey, hostNonce, clientNonce)
        var remainder = 0
        for (byte in digest) {
            remainder = (remainder * 256 + (byte.toInt() and 0xff)) % 1_000_000
        }
        return SasCode(String.format(Locale.US, "%06d", remainder))
    }

    private fun sha256(vararg parts: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        parts.forEach(digest::update)
        return digest.digest()
    }
}

File: mac/Sources/EdgeLinkKit/Pairing.swift
import CryptoKit
import Foundation

public struct SASCode: Equatable, Sendable {
    public let numeric: String
    public let display: String

    public init(numeric: String) {
        self.numeric = numeric
        self.display = "\(numeric.prefix(3)) \(numeric.suffix(3))"
    }
}

public enum Pairing {
    public static func commitment(hostPublicKey: Data, hostNonce: Data) -> Data {
        sha256(hostPublicKey + hostNonce)
    }

    public static func sas(hostPublicKey: Data, clientPublicKey: Data, hostNonce: Data, clientNonce: Data) -> SASCode {
        let digest = sha256(hostPublicKey + clientPublicKey + hostNonce + clientNonce)
        var remainder = 0
        for byte in digest {
            remainder = (remainder * 256 + Int(byte)) % 1_000_000
        }
        return SASCode(numeric: String(format: "%06d", remainder))
    }

    private static func sha256(_ data: Data) -> Data {
        Data(SHA256.hash(data: data))
    }
}
